File: java/com/google/copybara/ReadConfigFromChangeWorkflow.java
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara;

import static com.google.common.base.Joiner.on;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.copybara.exception.ValidationException.checkCondition;
import static java.lang.String.format;

import com.google.common.base.MoreObjects;
import com.google.common.base.Preconditions;
import com.google.copybara.Destination.Writer;
import com.google.copybara.Origin.Reader;
import com.google.copybara.WorkflowRunHelper.ChangeMigrator;
import com.google.copybara.config.Config;
import com.google.copybara.config.ConfigValidator;
import com.google.copybara.config.Migration;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.monitor.EventMonitor.ChangeMigrationFinishedEvent;
import com.google.copybara.util.Glob;
import java.nio.file.Path;
import java.util.List;
import java.util.Set;
import java.util.function.Consumer;
import java.util.logging.Logger;
import javax.annotation.Nullable;

/**
 * An extension of {@link Workflow} that is capable of reloading itself, reading the configuration
 * from the origin location provided.
 *
 * <p>How it works is with the method
 * {@link Workflow#newRunHelper(Path, Revision, String, Consumer<ChangeMigrationFinishedEvent>)} and
 * {@link WorkflowRunHelper}. The core implementation returns a regular run helper that always
 * returns {@code this} for any changes, which means that no config is read and the workflow remains
 * immutable.
 *
 * <p>The service uses this implementation to provide a {@link ReloadingRunHelper} that is capable
 * of reading the configuration for the current change being migrated, perform some security and
 * validation checks, and provide a new run helper instance.
 */
public class ReadConfigFromChangeWorkflow<O extends Revision, D extends Revision>
    extends Workflow<O, D> {
  private final Logger logger = Logger.getLogger(this.getClass().getName());

  private final ConfigLoader configLoader;
  private final ConfigValidator configValidator;

  ReadConfigFromChangeWorkflow(Workflow<O, D> workflow, Options options,
      ConfigLoader configLoader, ConfigValidator configValidator) {
    super(
        workflow.getName(),
        workflow.getDescription(),
        workflow.getOrigin(),
        workflow.getDestination(),
        workflow.getAuthoring(),
        workflow.getTransformation(),
        workflow.getLastRevisionFlag(),
        workflow.isInitHistory(),
        options.get(GeneralOptions.class),
        workflow.getOriginFiles(),
        workflow.getDestinationFiles(),
        workflow.getMode(),
        workflow.getWorkflowOptions(),
        workflow.getReverseTransformForCheck(),
        workflow.getReversibleCheckIgnoreFiles(),
        workflow.isAskForConfirmation(),
        workflow.getMainConfigFile(),
        workflow.getAllConfigFiles(),
        workflow.isDryRunMode(),
        workflow.isCheckLastRevState(),
        workflow.getAfterMigrationActions(),
        workflow.getAfterAllMigrationActions(),
        workflow.getChangeIdentity(),
        workflow.isSetRevId(),
        workflow.isSmartPrune(),
        workflow.isMigrateNoopChanges(),
        workflow.customRevId(),
        workflow.isCheckout());
    this.configLoader = checkNotNull(configLoader, "configLoaderProvider");
    this.configValidator = checkNotNull(configValidator, "configValidator");
  }

  @Override
  protected WorkflowRunHelper<O, D> newRunHelper(Path workdir, O resolvedRef,
      String rawSourceRef, Consumer<ChangeMigrationFinishedEvent> migrationFinishedMonitor)
      throws ValidationException {
    Reader<O> reader = this.getOrigin().newReader(this.getOriginFiles(), this.getAuthoring());
    return new ReloadingRunHelper(
        this,
        getName(),
        workdir,
        resolvedRef,
        createWriter(resolvedRef),
        reader,
        rawSourceRef,
        migrationFinishedMonitor);
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this).toString();
  }

  /**
   * A {@link WorkflowRunHelper} that reloads itself based on the change being imported, loading
   * the configuration from the origin, after performing security and validation checks.
   */
  private class ReloadingRunHelper extends WorkflowRunHelper<O, D> {
    private final Workflow<O, D> workflow;
    private final String workflowName;

    private ReloadingRunHelper(
        Workflow<O, D> workflow,
        String workflowName,
        Path workdir,
        O resolvedRef,
        Writer<D> writer,
        Reader<O> originReader,
        @Nullable String rawSourceRef,
        Consumer<ChangeMigrationFinishedEvent> migrationFinishedMonitor) {

      super(workflow, workdir, resolvedRef, originReader, writer, rawSourceRef,
          migrationFinishedMonitor);
      this.workflow = workflow;
      this.workflowName = checkNotNull(workflowName, "workflowName");
    }

    @Override
    ChangeMigrator<O, D> getMigratorForChangeAndWriter(Change<?> change, Writer<D> writer)
        throws ValidationException, RepoException {
      checkNotNull(change);

      logger.info(format("Loading configuration for change '%s %s'",
          change.getRef(), change.firstLineMessage()));

      Config config = ReadConfigFromChangeWorkflow.this.configLoader.
          loadForRevision(getConsole(), change.getRevision());
      // The service config validator already checks that the configuration matches the registry,
      // checking that the origin and destination haven't changed.
      List<String> errors =
          configValidator
              .validate(config, workflowName)
              .getErrors();
      checkCondition(errors.isEmpty(), "Invalid configuration [ref '%s': %s ]: '%s': \n%s",
          change.getRef(), configLoader.location(), workflowName, on('\n').join(errors));

      Migration migration = config.getMigration(workflowName);
      checkCondition(migration instanceof Workflow,
          "Invalid configuration [ref '%s': %s ]: '%s' is not a workflow", change.getRef(),
          configLoader.location(), workflowName);
      @SuppressWarnings("unchecked")
      Workflow<O, D> workflowForChange = (Workflow<O, D>) migration;
      Reader<O> newReader = workflowForChange
          .getOrigin()
          .newReader(workflowForChange.getOriginFiles(), workflowForChange.getAuthoring());
      return new ReloadingChangeMigrator<>(
          workflow,
          workflowForChange,
          getWorkdir(),
          newReader,
          writer,
          getResolvedRef(),
          rawSourceRef,
          getMigrationFinishedMonitor());
    }
  }

  private static final class
  ReloadingChangeMigrator<O extends Revision, D extends Revision> extends ChangeMigrator<O, D> {

    private final Workflow<O, D> changeWorkflow;

    ReloadingChangeMigrator(Workflow<O, D> headWorkflow, Workflow<O, D> changeWorkflow,
        Path workdir, Reader<O> reader,
        Writer<D> writer, O resolvedRef, @Nullable String rawSourceRef,
        Consumer<ChangeMigrationFinishedEvent> migrationFinishedMonitor) {
      super(headWorkflow, workdir, reader, writer, resolvedRef, rawSourceRef,
          migrationFinishedMonitor);
      this.changeWorkflow = Preconditions.checkNotNull(changeWorkflow);
    }

    @Override
    protected Set<String> getConfigFiles() {
      return changeWorkflow.configPaths();
    }

    @Override
    protected Glob getOriginFiles() {
      return changeWorkflow.getOriginFiles();
    }

    @Override
    protected Glob getDestinationFiles() {
      return changeWorkflow.getDestinationFiles();
    }

    @Override
    protected Transformation getTransformation() {
      return changeWorkflow.getTransformation();
    }

    @Override
    @Nullable
    protected Transformation getReverseTransformForCheck() {
      return changeWorkflow.getReverseTransformForCheck();
    }

    @Override
    protected Glob getReversibleCheckIgnoreFiles() {
      return changeWorkflow.getReversibleCheckIgnoreFiles();
    }

  }
}

File: src/Copybara.Core/ReadConfigFromChangeWorkflow.cs
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Globalization;
using Copybara.Common;
using Copybara.Config;
using Copybara.Exceptions;
using Copybara.Revision;
using ChangeMigrationFinishedEvent = Copybara.Monitor.IEventMonitor.ChangeMigrationFinishedEvent;

namespace Copybara;

/// <summary>
/// An extension of <see cref="Workflow{O,D}"/> that is capable of reloading itself, reading the
/// configuration from the origin location provided.
///
/// <para>The core implementation returns a regular run helper that always returns <c>this</c> for any
/// changes, which means that no config is read and the workflow remains immutable. This
/// implementation provides a <c>ReloadingRunHelper</c> that reads the configuration for the current
/// change being migrated, performs security and validation checks, and provides a new run helper
/// instance.</para>
/// </summary>
/// <typeparam name="O">Origin revision type.</typeparam>
/// <typeparam name="D">Destination revision type.</typeparam>
public class ReadConfigFromChangeWorkflow<O, D> : Workflow<O, D>
    where O : class, IRevision
    where D : class, IRevision
{
    private readonly ConfigLoader _configLoader;
    private readonly ConfigValidator _configValidator;

    internal ReadConfigFromChangeWorkflow(
        Workflow<O, D> workflow,
        Options options,
        ConfigLoader configLoader,
        ConfigValidator configValidator)
        : base(
            workflow.GetName(),
            workflow.GetDescription(),
            workflow.GetOrigin(),
            workflow.GetDestination(),
            workflow.GetAuthoring(),
            workflow.GetTransformation(),
            workflow.GetLastRevisionFlag(),
            workflow.IsInitHistory(),
            options.Get<GeneralOptions>(),
            workflow.GetOriginFiles(),
            workflow.GetDestinationFiles(),
            workflow.GetMode(),
            workflow.GetWorkflowOptions(),
            workflow.GetReverseTransformForCheck(),
            workflow.GetReversibleCheckIgnoreFiles(),
            workflow.IsAskForConfirmation(),
            workflow.GetMainConfigFile(),
            workflow.GetAllConfigFiles(),
            workflow.IsDryRunMode(),
            workflow.IsCheckLastRevState(),
            workflow.GetAfterMigrationActions().ToImmutableArray(),
            workflow.GetAfterAllMigrationActions().ToImmutableArray(),
            workflow.GetChangeIdentity(),
            workflow.IsSetRevId(),
            workflow.IsSmartPrune(),
            workflow.GetMergeImport(),
            workflow.GetAutoPatchfileConfiguration(),
            workflow.AfterMergeTransformations,
            workflow.IsMigrateNoopChanges(),
            workflow.CustomRevId(),
            workflow.IsCheckout(),
            workflow.GetConsistencyFileConfig(),
            workflow.GetExpectedFixedRef(),
            workflow.GetPinnedFixedRef(),
            workflow.GetDefinitionStack(),
            workflow.GetDefinitionStackLocals())
    {
        _configLoader = Preconditions.CheckNotNull(configLoader, "configLoaderProvider");
        _configValidator = Preconditions.CheckNotNull(configValidator, "configValidator");
    }

    public override WorkflowRunHelper<O, D> NewRunHelper(
        string workdir,
        O resolvedRef,
        string? rawSourceRef,
        Action<ChangeMigrationFinishedEvent> migrationFinishedMonitor)
    {
        IOrigin<O>.IReader<O> reader = GetOrigin().NewReader(GetOriginFiles(), GetAuthoring());
        return new ReloadingRunHelper(
            this,
            this,
            GetName(),
            workdir,
            resolvedRef,
            CreateWriter(resolvedRef),
            reader,
            rawSourceRef,
            migrationFinishedMonitor);
    }

    public override string ToString() => "ReadConfigFromChangeWorkflow{}";

    /// <summary>
    /// A <see cref="WorkflowRunHelper{O,D}"/> that reloads itself based on the change being imported,
    /// loading the configuration from the origin, after performing security and validation checks.
    /// </summary>
    private sealed class ReloadingRunHelper : WorkflowRunHelper<O, D>
    {
        private readonly ReadConfigFromChangeWorkflow<O, D> _outer;
        private readonly Workflow<O, D> _workflow;
        private readonly string _workflowName;

        internal ReloadingRunHelper(
            ReadConfigFromChangeWorkflow<O, D> outer,
            Workflow<O, D> workflow,
            string workflowName,
            string workdir,
            O resolvedRef,
            IDestination<D>.IWriter<D> writer,
            IOrigin<O>.IReader<O> originReader,
            string? rawSourceRef,
            Action<ChangeMigrationFinishedEvent> migrationFinishedMonitor)
            : base(workflow, workdir, resolvedRef, originReader, writer, rawSourceRef,
                migrationFinishedMonitor)
        {
            _outer = outer;
            _workflow = workflow;
            _workflowName = Preconditions.CheckNotNull(workflowName, "workflowName");
        }

        internal override ChangeMigrator<O, D> GetMigratorForChangeAndWriter(
            Change<O> change, IDestination<D>.IWriter<D> writer)
        {
            Preconditions.CheckNotNull(change);

            // logger.info("Loading configuration for change '%s %s'", change.getRef(), ...);

            Config.Config config = _outer._configLoader.LoadForRevision(
                GetConsole(), change.GetRevision());
            // The service config validator already checks that the configuration matches the
            // registry, checking that the origin and destination haven't changed.
            IReadOnlyList<string> errors = _outer._configValidator
                .Validate(config, _workflowName)
                .GetErrors();
            ValidationException.CheckCondition(
                errors.Count == 0,
                "Invalid configuration [ref '%s': %s ]: '%s': \n%s",
                change.Ref,
                _outer._configLoader.Location(),
                _workflowName,
                string.Join("\n", errors));

            IMigration migration = config.GetMigration(_workflowName);
            ValidationException.CheckCondition(
                migration is Workflow<O, D>,
                "Invalid configuration [ref '%s': %s ]: '%s' is not a workflow",
                change.Ref,
                _outer._configLoader.Location(),
                _workflowName);
            var workflowForChange = (Workflow<O, D>)migration;
            IOrigin<O>.IReader<O> newReader = workflowForChange.GetOrigin()
                .NewReader(workflowForChange.GetOriginFiles(), workflowForChange.GetAuthoring());
            return new ReloadingChangeMigrator(
                _workflow,
                workflowForChange,
                GetWorkdir(),
                newReader,
                writer,
                GetResolvedRef(),
                RawSourceRef,
                GetMigrationFinishedMonitor());
        }
    }

    private sealed class ReloadingChangeMigrator : ChangeMigrator<O, D>
    {
        private readonly Workflow<O, D> _changeWorkflow;

        internal ReloadingChangeMigrator(
            Workflow<O, D> headWorkflow,
            Workflow<O, D> changeWorkflow,
            string workdir,
            IOrigin<O>.IReader<O> reader,
            IDestination<D>.IWriter<D> writer,
            O resolvedRef,
            string? rawSourceRef,
            Action<ChangeMigrationFinishedEvent> migrationFinishedMonitor)
            : base(headWorkflow, workdir, reader, writer, resolvedRef, rawSourceRef,
                migrationFinishedMonitor)
        {
            _changeWorkflow = Preconditions.CheckNotNull(changeWorkflow);
        }

        protected override Workflow<O, D> GetWorkflow() => _changeWorkflow;
    }
}
